File: src/main/java/org/fountainmc/api/chat/ComponentValueSerializer.java
package org.fountainmc.api.chat;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import org.fountainmc.api.chat.values.ComponentValue;
import org.fountainmc.api.chat.values.Text;
import org.fountainmc.api.chat.values.Translatable;

public class ComponentValueSerializer implements JsonSerializer<ComponentValue>, JsonDeserializer<ComponentValue> {

    @Override
    public ComponentValue deserialize(JsonElement element, Type type, JsonDeserializationContext context)
            throws JsonParseException {
        JsonObject object = element.getAsJsonObject();
        ComponentValue value = null;
        if (object.has("text")) {
            value = Text.of(object.get("text").getAsString());
        } else if (object.has("translate")) {
            Translatable translatable = Translatable.withId(object.get("translate").getAsString());
            if (object.has("with")) {
                List<String> with = new ArrayList<>();
                for (JsonElement a : object.getAsJsonArray("with")) {
                    with.add(a.getAsString());
                }
                translatable = translatable.withSubstitutions(with);
            }
            value = translatable;
        }

        if (value == null) {
            throw new JsonParseException("Invalid value");
        }

        return value;
    }

    @Override
    public JsonElement serialize(ComponentValue value, Type type, JsonSerializationContext context) {
        JsonObject object = new JsonObject();
        if (value instanceof Text) {
            object.addProperty("text", ((Text) value).getText());
        } else if (value instanceof Translatable) {
            object.addProperty("translate", ((Translatable) value).getMessage());
            List<String> with = ((Translatable) value).getSubstitutions();
            if (!with.isEmpty()) {
                JsonArray array = new JsonArray();
                with.forEach(array::add);
                object.add("with", array);
            }
        } else {
            throw new RuntimeException("Unable to serialize component value");
        }
        return object;
    }

}

File: src/main/kotlin/org/fountainmc/api/chat/ComponentValueSerializer.kt
package org.fountainmc.api.chat

import java.lang.reflect.Type
import java.util.ArrayList

import com.google.gson.JsonArray
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer

import org.fountainmc.api.chat.values.ComponentValue
import org.fountainmc.api.chat.values.Text
import org.fountainmc.api.chat.values.Translatable

class ComponentValueSerializer : JsonSerializer<ComponentValue>, JsonDeserializer<ComponentValue> {

    @Throws(JsonParseException::class)
    override fun deserialize(element: JsonElement, type: Type, context: JsonDeserializationContext): ComponentValue {
        val obj = element.asJsonObject
        var value: ComponentValue? = null
        if (obj.has("text")) {
            value = Text.of(obj.get("text").asString)
        } else if (obj.has("translate")) {
            var translatable = Translatable.withId(obj.get("translate").asString)
            if (obj.has("with")) {
                val with = ArrayList<String>()
                for (a in obj.getAsJsonArray("with")) {
                    with.add(a.asString)
                }
                translatable = translatable.withSubstitutions(with)
            }
            value = translatable
        }

        if (value == null) {
            throw JsonParseException("Invalid value")
        }

        return value
    }

    override fun serialize(value: ComponentValue, type: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()
        if (value is Text) {
            obj.addProperty("text", value.text)
        } else if (value is Translatable) {
            obj.addProperty("translate", value.message)
            val with = value.substitutions
            if (!with.isEmpty()) {
                val array = JsonArray()
                with.forEach { array.add(it) }
                obj.add("with", array)
            }
        } else {
            throw RuntimeException("Unable to serialize component value")
        }
        return obj
    }

}
